### PR TITLE
[JENKINS-16316] - direct unit test for the property

### PR DESCRIPTION
**Pull request by oleg-nenashev**
It's a version of #63 with the proper commit author specification

@recena @lanwen

### CODE REVIEW

File: src/test/java/org/jenkinsci/plugins/envinject/GlobalPropertiesTest.java [hunk 1]
```diff
@@ -91,4 +100,63 @@ public void testGlobalPropertiesSetWORKSPACE() throws Exception {
 
         assertEquals(result_testGlobalVariableName, result_testJobVariableName);
     }
+
+    @Test
+    @Issue("JENKINS-16316")
+    public void testChangeOfGlobalPropertyGetsRecognizedWhenWithoutJobPropertyAndRunOnSlaves() throws Exception {
```
Comment (lanwen) [inline src/test/java/org/jenkinsci/plugins/envinject/GlobalPropertiesTest.java]: please avoid such long tests. Test should not be longer than 10-20 lines. Or it does something wrong

File: src/test/java/org/jenkinsci/plugins/envinject/GlobalPropertiesTest.java [hunk 2]
```diff
@@ -91,4 +100,63 @@ public void testGlobalPropertiesSetWORKSPACE() throws Exception {
...
+        FreeStyleProject project = jenkins.createFreeStyleProject();
+
+        final String testVariableName = "TESTVAR";
```
Comment (lanwen) [inline src/test/java/org/jenkinsci/plugins/envinject/GlobalPropertiesTest.java]: this should be moved to const declaration

File: src/test/java/org/jenkinsci/plugins/envinject/GlobalPropertiesTest.java [hunk 3]
```diff
@@ -91,4 +100,63 @@ public void testGlobalPropertiesSetWORKSPACE() throws Exception {
...
+        final String testVariableValue = "value1";
+        final String testVariableValueAfterChange = "value2";
+
+        //A global node property TESTVAR
+        DescribableList<NodeProperty<?>, NodePropertyDescriptor> globalNodeProperties = jenkins.getInstance()
```
Comment (lanwen) [inline src/test/java/org/jenkinsci/plugins/envinject/GlobalPropertiesTest.java]: creation of the global node property should be in separate method

File: src/test/java/org/jenkinsci/plugins/envinject/GlobalPropertiesTest.java [hunk 4]
```diff
@@ -91,4 +100,63 @@ public void testGlobalPropertiesSetWORKSPACE() throws Exception {
...
+
+        Node slaveNode = jenkins.createOnlineSlave();
+        project.setAssignedNode(slaveNode);
+        project.getBuildersList().add(new Shell("echo \"TESTVAR=$TESTVAR\""));
+
+        // we do NOT add a jobProperty - we want to test "default" behaviour of jenkins with installed envinject plugin
+        // so we run the build right away
+        FreeStyleBuild build = project.scheduleBuild2(0).get();
+        assertEquals(Result.SUCCESS, build.getResult());
+        assertEquals(slaveNode.getNodeName(), build.getBuiltOn().getNodeName());
```
Comment (lanwen) [inline src/test/java/org/jenkinsci/plugins/envinject/GlobalPropertiesTest.java]: first case ends. Nothing should be after this line

File: src/test/java/org/jenkinsci/plugins/envinject/GlobalPropertiesTest.java [hunk 5]
```diff
@@ -91,4 +100,63 @@ public void testGlobalPropertiesSetWORKSPACE() throws Exception {
...
+
+        // assert correct injection of testVariable #1
```
Comment (lanwen) [inline src/test/java/org/jenkinsci/plugins/envinject/GlobalPropertiesTest.java]: second case starts

File: src/test/java/org/jenkinsci/plugins/envinject/GlobalPropertiesTest.java [hunk 6]
```diff
@@ -91,4 +100,63 @@ public void testGlobalPropertiesSetWORKSPACE() throws Exception {
...
+        org.jenkinsci.lib.envinject.EnvInjectAction action = build.getAction(org.jenkinsci.lib.envinject.EnvInjectAction.class);
+        Map<String, String> envVars = action.getEnvMap();
+        String actualTestVariableValueInBuild = envVars.get(testVariableName);
+        assertNotNull("actual testVariableValue is null", actualTestVariableValueInBuild);
+        assertEquals(testVariableValue, actualTestVariableValueInBuild);
+
+        Set<Entry<String, String>> beforeChange = jenkins.getInstance().getComputer(slaveNode.getNodeName())
```
Comment (lanwen) [inline src/test/java/org/jenkinsci/plugins/envinject/GlobalPropertiesTest.java]: third case start

File: src/test/java/org/jenkinsci/plugins/envinject/GlobalPropertiesTest.java [hunk 7]
```diff
@@ -91,4 +100,63 @@ public void testGlobalPropertiesSetWORKSPACE() throws Exception {
...
+                .getEnvironment()
+                .entrySet();
+
+        // now we change the global property variable value...
+        testVarEntry = new EnvironmentVariablesNodeProperty.Entry(testVariableName, testVariableValueAfterChange);
+        jenkins.getInstance().getGlobalNodeProperties().add(new EnvironmentVariablesNodeProperty(testVarEntry));
+
+        Set<Entry<String, String>> afterChange = jenkins.getInstance().getComputer(slaveNode.getNodeName()).getEnvironment()
```
Comment (lanwen) [inline src/test/java/org/jenkinsci/plugins/envinject/GlobalPropertiesTest.java]: one more testcase

**Comment by oleg-nenashev** [comment]
Retriggering the build

**Comment by oleg-nenashev** [comment]
Hmm, seems to be a kind of regression (or unfixed issue)

**Pull request closed without merging**